File: ocaml/AC19.ml
type category = X | M | A | S

let category_of_char c =
  match c with
  | 'x' -> X
  | 'm' -> M
  | 'a' -> A
  | 's' -> S
  | _ -> failwith "parse_category"

type comparison = LT | GT

let comparison_of_char c =
  match c with
  | '<' -> LT
  | '>' -> GT
  | _ -> failwith "comparison_of_char"

type answer =
  | Accept
  | Reject

type action =
  | Send of string
  | Answer of answer

let action_of_string s =
  match s with
  | "A" -> Answer Accept
  | "R" -> Answer Reject
  | _ -> Send s

type rule = {
    category : category;
    comparison : comparison;
    value : int;
    action : action;
  }

let rule_of_string s =
  Scanf.sscanf s "%c%c%d:%s" (fun category comparison value action ->
      { category = category_of_char category;
        comparison = comparison_of_char comparison;
        value;
        action = action_of_string action;
      })

type workflow = {
    rules : rule list;
    default : action;
  }

let workflow_of_string s =
  Scanf.sscanf s "%s@{%s@}" (fun name rule_str ->
      let rules = String.split_on_char ',' rule_str in
      let default, rev_rules =
        match List.rev rules with
        | [] -> assert false
        | default :: rev_rules -> default, rev_rules in
      let rules = List.rev_map rule_of_string rev_rules in
      name, { rules; default = action_of_string default }
    )

type 'a part = {
    x : 'a;
    m : 'a;
    a : 'a;
    s : 'a;
  }

let part_of_string s =
  Scanf.sscanf s "{x=%d,m=%d,a=%d,s=%d}"
    (fun x m a s -> { x; m; a; s })

let sum_part { x; m; a; s } =
  x + m + a + s

module StringMap = Map.Make (String)

let get_value { x; m; a; s } category =
  match category with
  | X -> x
  | M -> m
  | A -> a
  | S -> s

let set_value part category value =
  match category with
  | X -> { part with x = value }
  | M -> { part with m = value }
  | A -> { part with a = value }
  | S -> { part with s = value }

let get_comparison comparison =
  match comparison with
  | LT -> ( < )
  | GT -> ( > )

let check_rule part rule =
  get_comparison rule.comparison (get_value part rule.category) rule.value

let find_action workflow part =
  match List.find_opt (check_rule part) workflow.rules with
  | None -> workflow.default
  | Some rule -> rule.action

let rec find_answer workflows workflow part =
  let workflow = StringMap.find workflow workflows in
  match find_action workflow part with
  | Send target -> find_answer workflows target part
  | Answer answer -> answer

let is_accepted workflows part =
  find_answer workflows "in" part = Accept

let split_range (low, high) comparison value =
  match comparison with
  | LT ->
     if high < value then
       Some (low, high), None
     else if low < value then
       Some (low, value - 1), Some (value, high)
     else
       None, Some (low, high)
  | GT ->
     if low > value then
       Some (low, high), None
     else if high > value then
       Some (value + 1, high), Some (low, value)
     else
       None, Some (low, high)

let count_range (low, high) =
  high - low + 1

let count_parts { x; m; a; s } =
  count_range x * count_range m * count_range a * count_range s

let rec count_accepted_ratings workflows workflow parts =
  let count_action action parts =
    match action with
    | Send workflow_name ->
       let workflow = StringMap.find workflow_name workflows in
       count_accepted_ratings workflows workflow parts
    | Answer Accept -> count_parts parts
    | Answer Reject -> 0 in
  let rec split_rule rules parts =
    match rules with
    | [] -> count_action workflow.default parts
    | hd :: tl ->
       let yes_parts, no_parts =
         split_range (get_value parts hd.category) hd.comparison hd.value in
       let yes_count =
         match yes_parts with
         | None -> 0
         | Some yes_parts ->
            count_action hd.action (set_value parts hd.category yes_parts) in
       let no_count =
         match no_parts with
         | None -> 0
         | Some no_parts ->
            split_rule tl (set_value parts hd.category no_parts) in
       yes_count + no_count in
  split_rule workflow.rules parts

let () =
  let lines = List.of_seq (Common.input_lines ()) in
  let workflows, parts = Common.break (( = ) "") lines in
  let workflows =
    StringMap.of_list (List.map workflow_of_string workflows) in
  let parts = List.map part_of_string (List.tl parts) in
  let accepted = List.filter (is_accepted workflows) parts in
  let result_part1 = List.fold_left ( + ) 0 (List.map sum_part accepted) in
  Format.eprintf "Part 1: %d@." result_part1;
  let result_part2 =
    count_accepted_ratings workflows (StringMap.find "in" workflows)
      { x = (1, 4000); m = (1, 4000); a = (1, 4000); s = (1, 4000) } in
  Format.eprintf "Part 2: %d@." result_part2

File: rust/src/bin/ac19.rs
use advent_of_code::{Error, Result, NameTable, Name};

#[derive(Clone, Copy)]
enum Category { X, M, A, S }

impl TryFrom<char> for Category {
    type Error = Error;

    fn try_from(c: char) -> Result<Self> {
        match c {
            'x' => Ok(Category::X),
            'm' => Ok(Category::M),
            'a' => Ok(Category::A),
            's' => Ok(Category::S),
            _ => Err(format!("Unknown category: {c}"))?,
        }
    }
}

#[derive(Clone, Copy)]
enum Comparison { LT, GT }

impl Comparison {
    fn compare<T: PartialOrd>(self, a: T, b: T) -> bool {
        match self {
            Comparison::LT => a < b,
            Comparison::GT => a > b,
        }
    }
}

impl TryFrom<char> for Comparison {
    type Error = Error;

    fn try_from(c: char) -> Result<Self> {
        match c {
            '<' => Ok(Comparison::LT),
            '>' => Ok(Comparison::GT),
            _ => Err(format!("Unknown comparison: {c}"))?,
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
enum Answer { Accept, Reject }

#[derive(Clone, Copy)]
enum Action {
    Send(Name),
    Answer(Answer)
}

impl Action {
    fn parse(
        table: &mut NameTable, s: &str
    ) -> Self {
        match s {
            "A" => Action::Answer(Answer::Accept),
            "R" => Action::Answer(Answer::Reject),
            _ => Action::Send(table.get(s)),
        }
    }
}

struct Rule {
    category: Category,
    comparison: Comparison,
    value: u64,
    action: Action,
}

impl Rule {
    fn parse(
        table: &mut NameTable, s: &str
    ) -> Result<Self> {
        let mut chars = s.chars();
        let category: Category =
            chars.next().ok_or("Missing category")?.try_into()?;
        let comparison: Comparison =
            chars.next().ok_or("Missing comparison")?.try_into()?;
        let (value_str, action_str) =
            chars.as_str().split_once(":").ok_or("Missing colon")?;
        let value = value_str.parse()?;
        Ok(Rule {
            category, comparison, value,
            action: Action::parse(table, action_str),
        })
    }

    fn check(&self, part: &Part<u64>) -> bool {
        self.comparison.compare(*part.get(self.category), self.value)
    }
}

struct Workflow {
    rules: Vec<Rule>,
    default: Action,
}

impl Workflow {
    fn parse(
        table: &mut NameTable, s: &str
    ) -> Result<(Name, Self)> {
        let (name_str, rule_str) = s.split_once('{').ok_or("Missing workflow")?;
        let name = table.get(name_str);
        let mut rule_chars = rule_str.chars();
        rule_chars.next_back();
        let rules_str: Vec<_> = rule_chars.as_str().split(',').collect();
        let mut rules_iter = rules_str.iter();
        let default_str = rules_iter.next_back().ok_or("Missing default")?;
        let default = Action::parse(table, default_str);
        let workflow = Workflow {
            rules: rules_iter.map(|s| Rule::parse(table, s)).
                collect::<Result<_>>()?,
            default
        };
        Ok((name, workflow))
    }

    fn get_action(&self, part: &Part<u64>) -> Action {
        match self.rules.iter().find(|rule| rule.check(part)) {
            None => self.default,
            Some(rule) => rule.action,
        }
    }
}

struct Workflows {
    map: std::collections::HashMap<Name, Workflow>,
}

impl Workflows {
    fn parse<'a>(
        table: &mut NameTable,
        it: &mut impl Iterator<Item = &'a String>,
    ) -> Result<Workflows> {
        Ok(Workflows { map:
            it.map(|line| Workflow::parse(table, &line)).collect::<Result<_>>()?
        })
    }

    fn get_answer(&self, mut name: Name, part: &Part<u64>) -> Result<Answer> {
        loop {
            let workflow = self.map.get(&name).ok_or("Unknown workflow")?;
            match workflow.get_action(&part) {
                Action::Send(target) => name = target,
                Action::Answer(answer) => return Ok(answer),
            }
        }
    }
}

#[derive(Clone)]
struct Part<T: Clone> {
    x: T,
    m: T,
    a: T,
    s: T,
}

impl TryFrom<&str> for Part<u64> {
    type Error = Error;

    fn try_from(s: &str) -> Result<Self> {
        let mut chars = s.chars();
        if chars.next() != Some('{') {
            Err("Missing {")?
        }
        if chars.next_back() != Some('}') {
            Err("Missing }")?
        }
        let components: Vec<_> = chars.as_str().split(',').collect();
        let (x, m, a, s) =
            match &components[..] {
                &[x, m, a, s] => (x, m, a, s),
                _ => Err("4 components expected")?,
            };
        Ok(Part {
            x: x.strip_prefix("x=").ok_or("x expected")?.parse()?,
            m: m.strip_prefix("m=").ok_or("m expected")?.parse()?,
            a: a.strip_prefix("a=").ok_or("a expected")?.parse()?,
            s: s.strip_prefix("s=").ok_or("s expected")?.parse()?,
        })
    }
}

impl Part<u64> {
    fn sum(&self) -> u64 {
        self.x + self.m + self.a + self.s
    }

    fn is_accepted(
        &self, table: &mut NameTable, workflows: &Workflows
    ) -> Result<bool> {
        Ok(workflows.get_answer(table.get("in"), self)? == Answer::Accept)
    }
}

impl<T: Clone> Part<T> {
    fn get(&self, c: Category) -> &T {
        match c {
            Category::X => &self.x,
            Category::M => &self.m,
            Category::A => &self.a,
            Category::S => &self.s,
        }
    }

    fn set(&mut self, c: Category, v: T) {
        match c {
            Category::X => self.x = v,
            Category::M => self.m = v,
            Category::A => self.a = v,
            Category::S => self.s = v,
        }
    }
}

#[derive(Clone, Copy)]
struct Range {
    low: u64,
    high: u64,
}

impl Range {
    fn split(
        self, comparison: Comparison, value: u64
    ) -> (Option<Range>, Option<Range>) {
        match comparison {
            Comparison::LT =>
                if self.high < value {
                    (Some(self), None)
                }
                else if self.low < value {
                    (Some(Range { low: self.low, high: value - 1 }),
                     Some(Range { low: value, high: self.high }))
                }
                else {
                    (None, Some(self))
                },
            Comparison::GT =>
                if self.low > value {
                    (Some(self), None)
                }
                else if self.high > value {
                    (Some(Range { low: value + 1, high: self.high }),
                     Some(Range { low: self.low, high: value }))
                }
                else {
                    (None, Some(self))
                }
        }
    }

    fn count(self) -> u64 {
        self.high - self.low + 1
    }
}

impl Part<Range> {
    fn count(&self) -> u64 {
        self.x.count() * self.m.count() * self.a.count() * self.s.count()
    }

    fn count_action(
        self, table: &mut NameTable, workflows: &Workflows, action: Action
    ) -> Result<u64> {
        match action {
            Action::Send(name) => self.count_accepted(table, workflows, name),
            Action::Answer(Answer::Accept) => Ok(self.count()),
            Action::Answer(Answer::Reject) => Ok(0),
        }
    }

    fn count_accepted(
        mut self, table: &mut NameTable, workflows: &Workflows, name: Name
    ) -> Result<u64> {
        let workflow = workflows.map.get(&name).ok_or("Unknown workflow")?;
        let mut sum = 0;
        for rule in &workflow.rules {
            let (yes_range, no_range) =
                self.get(rule.category).split(rule.comparison, rule.value);
            match yes_range {
                None => (),
                Some(yes_range) => {
                    let mut yes_part = self.clone();
                    yes_part.set(rule.category, yes_range);
                    sum += yes_part.count_action(table, workflows, rule.action)?;
                }
            };
            match no_range {
                None => return Ok(sum),
                Some(no_parts) => self.set(rule.category, no_parts)
            };
        }
        Ok(sum + self.count_action(table, workflows, workflow.default)?)
    }
}

fn main() -> Result<()> {
    let lines: Vec<_> = std::io::stdin().lines()
        .collect::<std::result::Result<_, _>>()?;
    let mut table = NameTable::new();
    let mut lines_iter = lines.iter();
    let mut workflow_lines =
        lines_iter.by_ref().take_while(|line| !line.is_empty());
    let workflows = Workflows::parse(&mut table, &mut workflow_lines)?;
    let parts: Vec<_> = lines_iter.map(|line| Part::try_from(line.as_str()))
        .collect::<Result<_>>()?;
    let accepted_vec: Vec<_> =
        parts.iter().map(|part|
          part.is_accepted(&mut table, &workflows).map(|b| (part, b))).
        collect::<Result<_>>()?;
    let result_part1: u64 = accepted_vec.into_iter()
        .filter_map(|(part, b)| b.then(|| part.sum())).sum();
    println!("Part 1: {result_part1}");
    let part2_range = Range { low: 1, high: 4000 };
    let initial = table.get("in");
    let result_part2: u64 =
        Part { x: part2_range, m: part2_range, a: part2_range, s: part2_range }
        .count_accepted(&mut table, &workflows, initial)?;
    println!("Part 2: {result_part2}");
    Ok(())
}
